core/src/test/java/org/elasticsearch/index/percolator/QueryMetadataServiceTests.java

```diff
@@ -0,0 +1,224 @@
+        assertThat(terms.get(3).bytes(), equalTo(termQuery4.getTerm().bytes()));
+    }
+
+    public void testExtractQueryMetadata_booleanQueryWithMustNot() {
+        BooleanQuery.Builder builder = new BooleanQuery.Builder();
+        TermQuery termQuery1 = new TermQuery(new Term("_field", "_term"));
+        builder.add(termQuery1, BooleanClause.Occur.MUST_NOT);
+        PhraseQuery phraseQuery = new PhraseQuery("_field", "_term1", "term2");
+        builder.add(phraseQuery, BooleanClause.Occur.SHOULD);
+
+        BooleanQuery booleanQuery = builder.build();
+        List<Term> terms = queryMetadataService.extractQueryMetadata(booleanQuery);
+        assertThat(terms.size(), equalTo(1));
+        assertThat(terms.get(0).field(), equalTo(phraseQuery.getTerms()[0].field()));
+        assertThat(terms.get(0).bytes(), equalTo(phraseQuery.getTerms()[0].bytes()));
+    }
+
+    public void testExtractQueryMetadata_constantScoreQuery() {
+        TermQuery termQuery1 = new TermQuery(new Term("_field", "_term"));
+        ConstantScoreQuery constantScoreQuery = new ConstantScoreQuery(termQuery1);
+        List<Term> terms = queryMetadataService.extractQueryMetadata(constantScoreQuery);
+        assertThat(terms.size(), equalTo(1));
+        assertThat(terms.get(0).field(), equalTo(termQuery1.getTerm().field()));
+        assertThat(terms.get(0).bytes(), equalTo(termQuery1.getTerm().bytes()));
+    }
+
+    public void testExtractQueryMetadata_boostQuery() {
+        TermQuery termQuery1 = new TermQuery(new Term("_field", "_term"));
+        BoostQuery constantScoreQuery = new BoostQuery(termQuery1, 1f);
+        List<Term> terms = queryMetadataService.extractQueryMetadata(constantScoreQuery);
+        assertThat(terms.size(), equalTo(1));
+        assertThat(terms.get(0).field(), equalTo(termQuery1.getTerm().field()));
+        assertThat(terms.get(0).bytes(), equalTo(termQuery1.getTerm().bytes()));
+    }
+
+    public void testExtractQueryMetadata_unsupportedQuery() {
+        TermRangeQuery termRangeQuery = new TermRangeQuery("_field", null, null, true, false);
+
+        try {
+            queryMetadataService.extractQueryMetadata(termRangeQuery);
+            fail("IllegalArgumentException expected");
+        } catch (IllegalArgumentException e) {
+            assertThat(e.getMessage(), equalTo("unsupported query"));
+        }
+
+        TermQuery termQuery1 = new TermQuery(new Term("_field", "_term"));
+        BooleanQuery.Builder builder = new BooleanQuery.Builder();;
+        builder.add(termQuery1, BooleanClause.Occur.SHOULD);
+        builder.add(termRangeQuery, BooleanClause.Occur.SHOULD);
+
+        try {
+            queryMetadataService.extractQueryMetadata(builder.build());
+            fail("IllegalArgumentException expected");
+        } catch (IllegalArgumentException e) {
+            assertThat(e.getMessage(), equalTo("unsupported query"));
+        }
+    }
+
+    public void testCreateQueryMetadataQuery() throws Exception {
```

Comment:
This test tests QMS#createMetadataQuery which just iterates over all fields in the in memory index and create a big boolean should query. It doesn't have the logic for unsupported queries that is in the #extractQueryMetadata(Query) method.